propagation/b3/grpc.go

```diff
@@ -0,0 +1,78 @@
+package b3
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc/metadata"
+
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/propagation"
+)
+
+// ExtractGRPC will extract a span.Context from the gRPC Request metadata if
+// found in B3 header format.
+func ExtractGRPC(md *metadata.MD) propagation.Extractor {
+	return func() (*model.SpanContext, error) {
+		var (
+			traceIDHeader      = GetGRPCHeader(md, TraceID)
+			spanIDHeader       = GetGRPCHeader(md, SpanID)
+			parentSpanIDHeader = GetGRPCHeader(md, ParentSpanID)
+			sampledHeader      = GetGRPCHeader(md, Sampled)
+			flagsHeader        = GetGRPCHeader(md, Flags)
+		)
+
+		return parseHeaders(
+			traceIDHeader, spanIDHeader, parentSpanIDHeader, sampledHeader,
+			flagsHeader,
+		)
+	}
+}
+
+// InjectGRPC will inject a span.Context into gRPC metadata.
+func InjectGRPC(md *metadata.MD) propagation.Injector {
+	return func(sc model.SpanContext) error {
+		if (model.SpanContext{}) == sc {
+			return ErrEmptyContext
+		}
+
+		if sc.Debug {
+			setGRPCHeader(md, Flags, "1")
+		} else if sc.Sampled != nil {
+			// Debug is encoded as X-B3-Flags: 1. Since Debug implies Sampled,
+			// so don't also send "X-B3-Sampled: 1".
+			if *sc.Sampled {
+				setGRPCHeader(md, Sampled, "1")
+			} else {
+				setGRPCHeader(md, Sampled, "0")
+			}
+		}
+
+		if !sc.TraceID.Empty() {
```

Comment:
scary this is possible! (trace and span id are required, so other side will reset the trace)

Reply:
From tracer implementation two situations are possible:

- Valid Identifiers exist (both TraceID and SpanID, optionally ParentSpanID) and optionally Sampled or Debug
- Only Debug flag exists (so we can instruct downstream to enforce sampling)